Introducing Identity for OpenSearch blog post

peternied opened the pull request:
### Description
Introducing Identity for OpenSearch blog post

### Issues Resolved
- Resolves https://github.com/opensearch-project/project-website/issues/1148 

### Check List
- [X] Commits are signed per the DCO using --signoff


By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

peternied commented:
@opensearch-project/security Could you take a look at how we are presenting identity more broadly to the community?  If anyone else wants to get an authorship credit happy to get additional contributors -  let me know!

peternied commented:
@jimishs @shanilpa Can you double-check that this is in alignment with the public discussion we want to start about these features?  Would love to get nitpicky feedback about the blog's content.

peternied commented on _posts/2022-12-05-Introducting-Identity.markdown:
```diff
@@ -0,0 +1,35 @@
+Actions have an existing protection model – resources of a plugin do not, they must be implemented by each plugin developer separately.  Being able to use shared systems for secure access and standard permissions schemes will make adding security features faster with fewer bugs.
+
+## How to learn more?
+Identity features are being built in a feature branch of OpenSearch, features/identity [4].  Roadmaps, documentation, findings, and functionality are in active development of that feature branch.  Beginning in December there will be a monthly check-in during the OpenSearch community meeting.
```
@krisfreedain We've discussed a series of presentations/participation in the community meetings that I've hand-waved into this doc.  What do you think of the proposal for a monthly update in the community meeting, or is there another format/audience that would better suit this discussion?

peternied replied:
@krisfreedain Any thoughts?

shanilpa commented:
@peternied Awesome job on this and can't wait to see what comes out of the community discussions! Here's some questions and maybe not so relevant feedback:

1. Does the phrase "All OpenSearch Systems" include OSDashboards? This feels heavily focused on identity as a backend construct - I know this is what enables the front end identity system but curious if we want to open it up a little more to invite some of those OS Dashboards identity discussions as well. "Who can access visualizations" "Who can share" etc.
2. I like the definitions in the footnotes however they link to a page with many definitions (some similar but not exact and this can be misinterpreted). Is out plan to work with the community to come to a common definition for these terms? If so I think that's something worth mentioning.
3. This is a language suggestion - the document is written well but uses some uncommon/complex words which might be hard for people who aren't that strong in English to understand. Would we be able to simplify the language and make it more accessible so we don't exclude some people from contributing to discussions cause they had a hard time understanding the doc?
4. What is a source of interaction? Seems like a central concept but I'm not sure I understand it looking through the document. - *this feedback might be because this is a common term and I might not be the audience for this doc
5. "Actions have an existing protection model" - what are these actions? What's an action? *again maybe a common term and I'm not really the audience for this doc

Hopefully this is the type of feedback you were looking for and makes sense 😅

peternied commented:
Great feedback @shanilpa 

>1. Does the phrase "All OpenSearch Systems" include OSDashboards? This feels heavily focused on identity as a backend construct - I know this is what enables the front end identity system but curious if we want to open it up a little more to invite some of those OS Dashboards identity discussions as well. "Who can access visualizations" "Who can share" etc.

Oh- good catch.  I am referring only to the back-end components in this doc - do you have a recommendation on how I could position the document to make that clear?

The front-end ecosystem can certainly use these identities through new APIs.  Features for relationship management/sharing scenarios are out of scope for what is planned in the roadmaps - but they could be developed in parallel with the existing security plugin features.

>2. I like the definitions in the footnotes however they link to a page with many definitions (some similar but not exact and this can be misinterpreted). Is out plan to work with the community to come to a common definition for these terms? If so I think that's something worth mentioning.

I'll reword to highlight these concepts to be more approachable

>3. This is a language suggestion - the document is written well but uses some uncommon/complex words which might be hard for people who aren't that strong in English to understand. Would we be able to simplify the language and make it more accessible so we don't exclude some people from contributing to discussions cause they had a hard time understanding the doc?

I'll do a pass to reword/simplify if you could make comments inline where you have specific suggestions that will make sure I don't miss anything.  I'm showing my bias toward complex language

>4. What is a source of interaction? Seems like a central concept but I'm not sure I understand it looking through the document. - *this feedback might be because this is a common term and I might not be the audience for this doc

The sources are that bulleted list - but this might be a way I could set up the conceptual model.  Ultimately everything is done by users, but sometimes the user does something directly via an API call, and other times they use a 3rd party system to make an API call.  IMO we should be able to treat those differently.  

>5. "Actions have an existing protection model" - what are these actions? What's an action? *again maybe a common term and I'm not really the audience for this doc

I'll expand as well and add a citation to the existing permissions documentation for those that want to dig in.

cliu123 commented on _posts/2022-12-05-Introducting-Identity.markdown:
```diff
@@ -0,0 +1,35 @@
+Identity features are being built in a feature branch of OpenSearch, features/identity [4].  Roadmaps, documentation, findings, and functionality are in active development of that feature branch.  Beginning in December there will be a monthly check-in during the OpenSearch community meeting.
+Feel free to add comments, create pull requests, raise issues and add the tag ‘identity’ to make sure we see it.
+
+Find me @peternied on github in the OpenSearch-Project.
```
Is this needed? If yes, would it be tag a team alias here?

cliu123 commented on _posts/2022-12-05-Introducting-Identity.markdown:
```diff
@@ -0,0 +1,35 @@
+Identity features are being built in a feature branch of OpenSearch, features/identity [4].  Roadmaps, documentation, findings, and functionality are in active development of that feature branch.  Beginning in December there will be a monthly check-in during the OpenSearch community meeting.
+Feel free to add comments, create pull requests, raise issues and add the tag ‘identity’ to make sure we see it.
+
+Find me @peternied on github in the OpenSearch-Project.
```
Is this needed? If yes, would it be better to tag a team alias here?

peternied replied:
This is a good callout, I'm not sure how we should be flagged down... I'll keep this open till we have an update.

aoguan1990 replied:
@peternied @cliu123 One option is to add link for the [OpenSearch Forum](https://forum.opensearch.org/c/feedback/6) to capture feedbacks from public.

pajuric commented:
Tentatively scheduled for publication on 12/6/2022.

cwperks commented on _posts/2022-12-05-Introducting-Identity.markdown:
```diff
@@ -0,0 +1,35 @@
+![image](https://user-images.githubusercontent.com/2754967/201424268-5687f162-4857-424b-96b8-ca8aff5f649a.png)
+
+## Why have identity in all OpenSearch systems?
+By knowing who is performing actions we can check if that user is Authorized [3].  Plugins and extensions can only do actions that have been permitted to their identity.  The existing security plugin does not have the ability to control what other plugins can do.
```
I am wondering if we should keep consistent with subject vs user as we often have debate on identity PRs around the topic too. 

To make the distinction clear do you think it makes sense to have a sentence of 2 somewhere in this post along the lines of: 

```User and Subject are often used interchangeably. For this blog post I will use the term Subject to represent a generalized user. For example, when a user queries OpenSearch, their subject will be associated with a list of permissions and evaluated before performing the query to determine if the action is authorized. In the model we are proposing, extensions will also be associated with a Subject. Administrators will be able to assign permissions to extensions to isolate how an extension can interact with OpenSearch.```

peternied replied:
This is a much better rewrite, unfortunately I pulled all writing about Subjects from the post, but we will get back to this in future examples

cwperks commented on _posts/2022-12-05-Introducting-Identity.markdown:
```diff
@@ -0,0 +1,35 @@
+Let’s permeate identity in every aspect of OpenSearch – enabling security by default - everywhere.
+
+## What are OpenSearch’s identities?
+In the literature, an identity is an Authenticated [1] Subject [2].  In the systems of OpenSearch, an identity embodies sources of interaction.
```
Thoughts on `In OpenSearch systems, interaction with a cluster should be clearly associated with an identity. Sources of interaction include: ...`

cwperks commented on _posts/2022-12-05-Introducting-Identity.markdown:
```diff
@@ -0,0 +1,35 @@
+
+Within the OpenSearch Project, there is a plugin Security that allows finer levels of access control providing a more satisfying answer.  Security uses the plugin architecture of OpenSearch that forces concepts around identity to be optional in core, plugins, and extensions components.
+
+Let’s permeate identity in every aspect of OpenSearch – enabling security by default - everywhere.
```
🙌

cwperks reviewed (commented):
Finished a first pass. Thank you for putting this together and making it public @peternied ! I am really looking forward to community input on this initiative.

cwillum commented on _posts/2022-12-05-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,40 @@
+
+OpenSearch handles critical data; how that data is permissioned and how a cluster is operated is a Big Deal™.  Consider the question "Who can delete all the indices in the cluster?", in the core of OpenSearch, the answer is whoever can make a request to the cluster.  
+
+Within the OpenSearch Project, the Security plugin allows finer levels of access control providing a more satisfying answer.  Security uses the plugin architecture of OpenSearch that forces concepts around identity to be optional in core, plugins, and extensions components.  Optional security is not enough for critical data, we will do to do better.
```
It might be helpful to include a straightforward definition of identity here. What is it, and how does it change OpenSearch's approach to security?

peternied replied:
Thanks for the great feedback and the offline session - I think I've restructured the content and delivery much more cleanly and properly focused on introducing these concepts and their values, love to know more about what you think.

cwillum commented on _posts/2022-12-05-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,40 @@
+# Introducing Identity for OpenSearch
+
+OpenSearch handles critical data; how that data is permissioned and how a cluster is operated is a Big Deal™.  Consider the question "Who can delete all the indices in the cluster?", in the core of OpenSearch, the answer is whoever can make a request to the cluster.  
```
I'm not sure what "Big Deal™" is. I googled this term and didn't get any hits that made sense.

peternied replied:
Dropped this kind of tone and language.

cwillum commented on _posts/2022-12-05-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,40 @@
+## Identity is more than a user
+In this blog post we use the term  *Subject* that includes users and non-person entities.  Following NIST's defintion Subjects are '[a]n individual, process, or device that causes information to flow among objects or change to the system state.'
+
+Among the OpenSearch-Project's scenario there is monitoring, data management, replication, machine learning, and notificatios.  The identities of thse systems and process need a common description as a Subject.
```
The term Subject is introduced here. But the term doesn't appear again in subsequent sections of the blog. Is it important to include it if it won't be used later on in this blog? If it is an integral part of understanding "identity" it would be good to draw a clearer connection between the two. Are they synonymous?

peternied replied:
Removed the term completely, its better suite in a more technical context

cwillum commented on _posts/2022-12-05-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,40 @@
+![image](https://user-images.githubusercontent.com/2754967/201424268-5687f162-4857-424b-96b8-ca8aff5f649a.png)
+
+## Why introduce identity into the core of OpenSearch?
+By knowing who is performing actions we can check if that user is Authorized [3].  Plugins and extensions can only do actions that have been permitted to their identity.  The existing security plugin does not have the ability to control what other plugins can do.
```
Where is the footnote for this: "Authorized [3]"? A separate file from this one?

peternied replied:
Heavily rewritten!

cwillum commented on _posts/2022-12-05-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,40 @@
+
+## Why introduce identity into the core of OpenSearch?
+By knowing who is performing actions we can check if that user is Authorized [3].  Plugins and extensions can only do actions that have been permitted to their identity.  The existing security plugin does not have the ability to control what other plugins can do.
+Protections of high-level tasks and low-level actions and resources be enforced OpenSearch-wide.  Plugins and extensions can access and depend on these systems to protect the features and functionality they bring to OpenSearch.
```
Somewhere here there needs to be a precise answer to the question being posed in the heading. Why introduce Identity? Something like, "Identity allows the Security plugin to protect the features and functionality that other plugins bring to OpenSearch."

cwillum commented on _posts/2022-12-05-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,40 @@
+By knowing who is performing actions we can check if that user is Authorized [3].  Plugins and extensions can only do actions that have been permitted to their identity.  The existing security plugin does not have the ability to control what other plugins can do.
+Protections of high-level tasks and low-level actions and resources be enforced OpenSearch-wide.  Plugins and extensions can access and depend on these systems to protect the features and functionality they bring to OpenSearch.
+
+## How does this benefit plugins/extensions?
```
This section covers ideas that can probably be rolled into the preceding section. The benefits seem related to the reasons why identity is being introduced in OpenSearch. Consider removing the heading and combining the content.

peternied replied:
I've restructured to handle this more much gently - great notes.

cwillum reviewed (commented):
Some general comments on the structure of the blog. I'll be available to clarify anything or discuss further, if it helps. Looking forward to seeing this rolled out.

DarshitChanpura commented on _posts/2022-12-05-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,40 @@
+![image](https://user-images.githubusercontent.com/2754967/201424268-5687f162-4857-424b-96b8-ca8aff5f649a.png)
+
+## Why introduce identity into the core of OpenSearch?
+By knowing who is performing actions we can check if that user is Authorized [3].  Plugins and extensions can only do actions that have been permitted to their identity.  The existing security plugin does not have the ability to control what other plugins can do.
```
Thoughts on replacing this bullet points like:
```
Identity is expected to cover current issues that are present in the OpenSearch project like:
* Security plugin doesn't have ability to control the permissions/access for other plugins
* Limited protection on tasks and actions OpenSearch wide
* Shared security model with central permission/access schemes
```

DarshitChanpura reviewed (commented):
@peternied Thank you for adding this blogpost. Here are my initial thoughts:

DarshitChanpura commented on _posts/2022-12-05-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,40 @@
+## What is the future of the Security Plugin?
+The Security Plugin has user identities and it also offers features such as encryption in transit, audit logging, and compliance features.  All these featuers are critical to the OpenSearch-Project and will continue to be available and supported.
+
+## How to learn more?
```
Can be re-worded as: `Interested in learning more?` Thoughts?

peternied commented on _posts/2022-12-05-Introducting-Identity.markdown:
```diff
@@ -0,0 +1,35 @@
+Actions have an existing protection model – resources of a plugin do not, they must be implemented by each plugin developer separately.  Being able to use shared systems for secure access and standard permissions schemes will make adding security features faster with fewer bugs.
+
+## How to learn more?
+Identity features are being built in a feature branch of OpenSearch, features/identity [4].  Roadmaps, documentation, findings, and functionality are in active development of that feature branch.  Beginning in December there will be a monthly check-in during the OpenSearch community meeting.
```
Reworked this, let me know if you think this flows better.

peternied commented on _posts/2022-12-05-Introducting-Identity.markdown:
```diff
@@ -0,0 +1,35 @@
+Protections of high-level tasks and low-level actions and resources be enforced OpenSearch-wide.  Plugins and extensions can access and depend on these systems to protect the features and functionality they bring to OpenSearch.
+
+## How does this benefit plugins/extensions?
+Actions have an existing protection model – resources of a plugin do not, they must be implemented by each plugin developer separately.  Being able to use shared systems for secure access and standard permissions schemes will make adding security features faster with fewer bugs.
```
I've refocused this doc to be much less technical and more user-centric, I think we will build on some of this content in more 'deep' dive posts or documents.

peternied reviewed (commented):
Thanks everyone for the great feedback, I've restructured the document trying to balance all of this into account.  Even if you don't see your individual line item updates, know that I've got like four more technical docs brewing that will use that feedback - so thank you!

Housekeeping wise - pushed this out to a much more achievable mid-January timeframe.

DarshitChanpura commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,44 @@
+
+As Identity impacts many facets of the OpenSearch ecosystem we would like your ideas and contributions to be engaged during the development process [link](https://github.com/opensearch-project/OpenSearch/issues/4514) and watch this [feature branch](https://github.com/opensearch-project/OpenSearch/pulls?q=is%3Apr+base%3Afeature%2Fidentity) or [label](https://github.com/opensearch-project/OpenSearc ...
+
+Expect more blog posts on Identity features and presentations during the community meetings.
```
`More updates to follow during community meetings.` Thoughts?

cwperks commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,44 @@
+
+### Enabling permissions checks 
+
+With any Using OpenSearch Dashboards your permissions you can start editing a Dashboard,and after hitting 'Save' seeing an error message "Dashboard ... was not saved. Error: Forbidden".  This scenario exists because Dashboards can not check if these commands are allowed or denied.
```
This is a good example of the limitation of the current security model, but I think it needs to be reframed. 

Does it make sense to present an analogy here? In the current system there is no way to ask the question "Does this user have permission to perform this action?", instead what happens is that the user tries to perform and action and is blocked after initiating an action. In this case it leads to less-than-desirable UX because UI actions cannot be disabled when its known that a user does not have high enough privileges.

peternied replied:
Restructured, let me know what you think

cwperks commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,44 @@
+
+### Plugins and Extensions restrictions
+
+Feeling confidant installating apps on mobile phone is in part because there are limits on what apps can do.  OpenSearch's plugins have had few limits making them risky for Administrators to install.
```
Also `installing`.

This is a great analogy and I am wondering if we should mention anything about OpenSearch being a platform. When moving to the new paradigm of extensions, cluster administrators need new tools to limit how extensions can interact with an OpenSearch cluster similar to how smartphone users can limit apps downloaded from an App Store.

peternied replied:
Fixed `installating`

peternied replied:
> OpenSearch being a platform

Good callout, not sure how to fit this in there, did you have a notation how to make this a more concrete callout?

cwperks commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,44 @@
+
+## Interested in being involved?
+
+These are bold changes and they are going to require doing things differently that has been done. We are in the process of defining and building the Identity systems into the core of open search. 
```
nit: Use `OpenSearch` consistently throughout.

cwperks commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,44 @@
+
+As the core OpenSearch project begins its shift away from a plugin model to a framework that utilizes extensions, those extensions, its legacy plugins, and the administrators who manage them will need mechanisms for controlling access that are more granular and able to cover a broader range of scenarios where effective access control is critical. We are creating/building out/developing a new suit ...
+
+The main objectives for Identity include:
```
🙌  I like this structure of listing high-level objects and diving into each one. These are my main takeaways when reading the post.

cwperks commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,44 @@
+
+OpenSearch's existing access control features included in the Security plugin let administrators apply access control to indexes and cluster actions so that users have the right permissions to do their work and the cluster is protected from unwanted activity. However, the features for access control that are currently in place do have certain limitations that can make it hard to use them with oth ...
+
+As the core OpenSearch project begins its shift away from a plugin model to a framework that utilizes extensions, those extensions, its legacy plugins, and the administrators who manage them will need mechanisms for controlling access that are more granular and able to cover a broader range of scenarios where effective access control is critical. We are creating/building out/developing a new suit ...
```
wdyt about linking to a blog post or this GH issue (https://github.com/opensearch-project/OpenSearch/issues/2447) on extensions here?

peternied replied:
I've added a link when I first mention extensions

cwperks reviewed (commented):
Thank you for the update @peternied! I took a first pass and focused more on high-level content.

I like the conciseness of the post. It conveys a lot of information for a short blog post. 

There are a couple of paragraphs that I think could use a bit of improvement when it comes to flow, but the core idea comes across in the post. 

Through the post, there are a lot of mentions of extensions and it left me wondering if maybe we can describe it as a `paradigm shift` for OpenSearch which spurs the need to rethink security for OpenSearch. Lately, I have been describing the work as `building a platform` and the security features necessary to operate the platform securely.

peternied commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,44 @@
+
+There is a blind spot for OpenSearch Administrators where after they make a change to permissions of a user or group it is unclear if those were applied as expected.
+
+Identity will add APIs that Dashboards, plugins, and extensions can use to check.  Administrators will be able to make requests to these same APIs to check permissions.
```
Reworked this, let me know if this flows better to you.

shanilpa commented:
Couple of suggestions: 

1. It would be good to provide a short description for key terms and what they mean. We should try to leave as little as possible to assumption especially around foundational terms or concepts. We could do this by defining each key term as simply and clearly as possible. I think appending this in footnotes works - doesn't have to be in the body. Some of the key terms I think we should include these definitions for.
- OpenSearch
- Extensions
- Plugins
- Permissions

2. “Feeling confident installing apps on mobile phone is in part because there are limits on what apps can do.” - I understand the analogy but we might need to rephrase it to read better. "When we install an app on our phones we don't need to think about the implications because we are confident there are limits to what an app can do"

peternied commented:
@opensearch-project/project-website I'd like to get this merged before the 16th, could a maintainer please review?

cwillum commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,44 @@
+
+### Security Isolation 
+
+Using the lowest level of privileges when any task is executed is a key to prevent mistakes or software defects from impacting the stability of an OpenSearch cluster. By enforcing protections everywhere inside OpenSearch, the blast radius of these incidents will be drastically reduced.
```
I was just looking over something that DC had commented on in another PR regarding the principle of least privilege. Is that the idea here? If yes, I suppose this should be reworded to "Using a minimal set of privileges to allow users to run tasks is key ...". If it's important to emphasize this idea, I could revise this to place a sharper point on the concept and make clear that it's a driver behind this objective.

peternied replied:
> I was just looking over something that DC had commented on in another PR regarding the principle of least privilege. Is that the idea here? If yes, [...]

Yes this is the case.  On the topic of who is creating these permissions, they will be managed internally by the OpenSearch team, so while administrators/users of a cluster will benefit, the only action they need to take is using the latest version.

cwillum replied:
Hold on, that's big ([permissions will be] managed internally by the OpenSearch team, so while administrators/users of a cluster will benefit, the only action they need to take is using the latest version). 
Are you adding that? At least, "the only action users need to take is to keep versions up to date.)

peternied commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,44 @@
+
+### Plugins and Extensions restrictions
+
+Feeling confident installing apps on mobile phone is in part because there are limits on what apps can do. OpenSearch's plugins have had few limits making them risky for Administrators to install.
```
You are a wordsmith of amazing caliber!

peternied commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,44 @@
+
+### Background tasks permissions
+
+Many tasks run in the background of an OpenSearch cluster such as deleting old indices, producing monitoring data, and generating reports. When these different tasks execute, each should only have minimal set of permissions it needs to protect against mistakes or unintended consequences.
```
Nicely done, I've trimmed off that last section about alerts, if someone build a log of success/failed tasks would be really cool :D

peternied commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,44 @@
+
+## Get Involved
+
+This is a paradigm shift, security features will be easier for OpenSearch developers to access and rely on. We are in the process of defining and building the Identity systems into the core of OpenSearch. 
```
> Q: Do you want to provide some instruction for ways to join the community meetings?

Good call, I'll add a link

peternied replied:
> Terms in this blog ...

Love this.  It looks like GitHub supports footnotes[^1], let me know what you think.

[^1]: https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/

cwillum replied:
That is pretty slick!

peternied commented:
I've updated the doc and resolved 90% of the open comments.

@shanilpa great idea to add footnotes, GitHub supports this 🤞 our website does too

DarshitChanpura reviewed (approved):
LGTM from security developer point of view

pajuric commented:
@peternied - Please add front matter to this blog. It should be included just below the title, author, and date at the top of the blog. Here's the content:

natebower commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,61 @@
+
+### Security isolation 
+
+Following the principle of least privilege, using a minimum number of permissions to run tasks is key to preventing execution errors or software flaws from having an impact on the stability of an OpenSearch cluster. Identity will launch features that can protect all of the separate elements in OpenSearch by isolating the potential impact a problem with one part may exert on another. As a conseque ...
```
```suggestion
Following the principle of least privilege, using a minimum number of permissions to run tasks is key to preventing execution errors or software flaws from impacting the stability of an OpenSearch cluster. Identity will launch features that can protect all of the separate elements in OpenSearch by isolating the potential impact a problem in one element may have on another element. This will reduce the spread of errors and keep them from affecting other areas of OpenSearch.
```

natebower commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,61 @@
+
+## Get involved
+
+"Identity presents a significant departure from the current security model. It does so to provide features that make it easier for OpenSearch developers to access and build assets and for administrators to carry out tasks in a stable environment. Over the last several months we’ve been busy defining new features and building out the tools to prepare for the integration of Security into core OpenS ...
```
"integration of the Security plugin into core OpenSearch"?

natebower commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,61 @@
+
+"Identity presents a significant departure from the current security model. It does so to provide features that make it easier for OpenSearch developers to access and build assets and for administrators to carry out tasks in a stable environment. Over the last several months we’ve been busy defining new features and building out the tools to prepare for the integration of Security into core OpenS ...
+
+And since we operate in the open source community, we’d like to learn about your ideas and benefit from your contributions as we make progress and move development along.
```
```suggestion
And since we operate in the open-source community, we’d like to learn about your ideas and benefit from your contributions as we make progress.
```

natebower replied:
How would we like to learn about their ideas and benefit from their contributions? What do we want them to do?

natebower commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,61 @@
+
+And since we operate in the open source community, we’d like to learn about your ideas and benefit from your contributions as we make progress and move development along.
+
+Keep an eye out for further blog posts on specific features for Identity, and join us for presentations during the community meetings. Furthermore, you can stay on top of development by visiting the following resources in the OpenSearch repository:
```
Should we include a link relevant to community meetings?

natebower commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,61 @@
+
+[^1]: OpenSearch — a community-driven and open source (Apache 2.0 licensed) search engine.
+
+[^2]: Permissions — rights and privileges granted to a user that allow the user to perform specified tasks within a software application or platform. Permissions are typically managed by administrators responsible for overseeing who gets access to different parts of the platform.
```
```suggestion
[^2]: Permissions – Rights and privileges granted to a user that allow the user to perform specified tasks within a software application or platform. Permissions are typically managed by administrators responsible for overseeing who can access different parts of the platform.
```

natebower commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,61 @@
+
+[^3]: Plugin — a small piece of software designed to customize a larger software application. Plugins do not modify the core functionality of the larger application, they only add to the core functionality.
+
+[^4]: Extension — a software program designed to enhance and extend functionality of a larger software application. Extensions do involve some integration with the core of the larger application to meet the aims of the extra functionality it introduces.
```
```suggestion
[^4]: Extension – A software program designed to enhance and extend the functionality of a larger software application. Extensions involve some integration with the core of the larger application to meet the aims of the extra functionality they introduce.
```

natebower commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,61 @@
+- [Feature branch for Identity](https://github.com/opensearch-project/OpenSearch/pulls?q=is%3Apr+base%3Afeature%2Fidentity)
+- [Current issues for Identity](https://github.com/opensearch-project/OpenSearch/issues?q=label%3AIdentity+)
+
+We look forward to your participation.
```
```suggestion
We look forward to your participation!
```

natebower reviewed (commented):
@peternied @cwillum Please see my comments and changes and let me know if you have any questions or would like to discuss. Thanks!

peternied commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,61 @@
+
+Identity’s main objectives aim to:
+
+- Provide mechanisms for OpenSearch, plugins, and extensions that can check permissions before attempting an action.
```
Rewrote as "Provide mechanisms for checking permissions in OpenSearch and in OpenSearch plugins and extensions before attempting an action." - OpenSearch internal components have never been able to conduct these checks so they are part of the equation too in my mind.

peternied commented on _posts/2023-01-16-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,61 @@
+# Introducing Identity for OpenSearch
```
Being that the total scope of this space is intentionally in flux shifting away from a product name such as "Identity" and instead referring to these components as 'identity and access controls'.  Thanks for calling this out!

peternied commented:
@natebower Thanks for all the great feedback; I've made all of those updates
@pajuric I have added the metadata at the top of the blog, an author profile and updated the date to the 18th.

natebower commented on _posts/2023-01-18-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,73 @@
+
+To work effectively, application developers need to know what users can and cannot do in OpenSearch based on the permissions assigned to them. Likewise, administrators responsible for setting up users and the permissions assigned to users require a reliable way to make sure the permissions are configured correctly. Given the degree of complexity that these configurations can reach in large system ...
+
+The approach to closing this gap is to provide a robust set of APIs that will allow these checks on permissions. Furthermore, these APIs should have the capacity to work not only in core OpenSearch but across all plugins and extensions as well.
```
```suggestion
Our approach to closing this gap is to provide a robust set of APIs that will allow these checks on permissions. Furthermore, these APIs should have the capacity to work not only in core OpenSearch but across all plugins and extensions as well.
```

natebower commented on _posts/2023-01-18-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,73 @@
+
+## Get involved
+
+These features present a significant departure from the current security model. It does so by providing features that make it easier for OpenSearch developers to access and build assets and for administrators to carry out tasks in a stable environment. Over the last several months, we’ve been busy defining new features and building tools to prepare for the integration of Security into core OpenSe ...
```
What does the uppercase instance of "Security" refer to? What is being integrated into core OpenSearch?

natebower commented on _posts/2023-01-18-Introducing-Identity.markdown:
```diff
@@ -0,0 +1,73 @@
+
+And since we operate in the open-source community, we’d like to learn about your ideas and benefit from your contributions as we make progress.
+
+Watch for further blog posts on specific identity and access control features, and join us for community meetings. Furthermore, you can stay informed of development by visiting the following resources in the OpenSearch repository:
```
I still think it would be a good idea to provide links to where the reader can submit their ideas or join a community meeting.

peternied replied:
Added a feedback link

natebower reviewed (commented):
@peternied Just a few more minor changes and you should be good to go.

natebower reviewed (dismissed):
LGTM

natebower commented:
@pajuric @krisfreedain Approved and ready to publish.

pajuric commented:
@peternied - Looks like the meta description and keywords fell off the post. Please confirm they are included before publishing. 
**Meta descrip:** Learn about OpenSearch Identity and how this suite of features provides users with comprehensive access control and creates a better defense-in-depth posture.
**Keywords:** access control, OpenSearch identity, OpenSearch security, least privilege access

peternied commented:
@pajuric I've added those fields, let me know if that captures what you are looking for

pajuric commented:
> @peternied - That resolves my comments. Thank you.

pajuric commented:
@krisfreedain - This is not a closed PR. It still needs to be merged and published.

krisfreedain reviewed (changes requested):
@peternied - two minor edits requested - thanks!

peternied commented:
@krisfreedain I've made those updates, thanks
